buffer/src/main/java/io/netty/buffer/api/Scope.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2021 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.buffer.api;
+
+import java.util.ArrayDeque;
+
+/**
+ * A scope is a convenient mechanism for capturing the life cycles of multiple reference counted objects. Once the scope
+ * is closed, all the added objects will also be closed in reverse insert order. That is, the most recently added
+ * object will be closed first.
+ * <p>
+ * Scopes can be reused. After a scope has been closed, new objects can be added to it, and they will be closed when the
+ * scope is closed again.
+ * <p>
+ * Objects will not be closed multiple times if the scope is closed multiple times, unless said objects are also added
+ * multiple times.
+ * <p>
+ * Note that scopes are not thread-safe. They are intended to be used from a single thread.
+ */
+public final class Scope implements AutoCloseable {
+    private final ArrayDeque<Resource<?>> deque = new ArrayDeque<>();
```

Comment:
If we always add/remove at the end, we can just use `ArrayList` which is faster for that purpose.

Reply:
I'm not convinced there's much of a performance difference to see here. Both have very simple implementations, but `ArrayDeque` has a nicer API.

Reply:
JIT might be smart enough but `ArrayDeque` is a circular queue which keeps the start offset while `ArrayList` always starts at the beginning of the internal array. Anyways, I'm not strong on this.

By the way, how about making `Scope` an interface? We could define a package-private `DefaultScope` that extends `ArrayList` or `ArrayDeque` so we can save one object allocation? It may be premature optimization, but making it an interface gives us more options for the future.

Reply:
Making it an interface makes sense.